src/ExpressionEvaluator/Core/Source/ExpressionCompiler/CustomTypeInfo.cs

```diff
@@ -1,37 +1,196 @@
+            ReadOnlyCollection<byte> payload,
+            out ReadOnlyCollection<byte> dynamicFlags,
+            out ReadOnlyCollection<string> tupleElementNames)
+        {
+            dynamicFlags = null;
+            tupleElementNames = null;
+
+            if ((payload == null) || (payloadTypeId != PayloadTypeId))
+            {
+                return;
+            }
+
+            int length = payload[0];
+            if (length > 0)
+            {
+                dynamicFlags = new ReadOnlyCollection<byte>(CopyBytes(payload, 1, length));
+            }
+
+            int start = length + 1;
+            if (start < payload.Count)
+            {
+                tupleElementNames = DecodeNames(payload, start);
+            }
+        }
+
+        private const char NameSeparator = '|';
+
+        private static ReadOnlyCollection<byte> EncodeNames(ReadOnlyCollection<string> names)
+        {
+            var pooledBuilder = PooledStringBuilder.GetInstance();
+            var builder = pooledBuilder.Builder;
+            foreach (var name in names)
+            {
+                if (builder.Length > 0)
+                {
+                    builder.Append(NameSeparator);
+                }
+                if (name != null)
+                {
+                    builder.Append(name);
+                }
+            }
+            var str = pooledBuilder.ToStringAndFree();
+            return new ReadOnlyCollection<byte>(Encoding.UTF8.GetBytes(str));
+        }
 
-        public CustomTypeInfo(Guid payloadTypeId, ReadOnlyCollection<byte> payload)
+        private static ReadOnlyCollection<string> DecodeNames(ReadOnlyCollection<byte> bytes, int start)
         {
-            this.PayloadTypeId = payloadTypeId;
-            this.Payload = payload;
+            int length = bytes.Count - start;
+            var array = CopyBytes(bytes, start, length);
+            var str = Encoding.UTF8.GetString(array, 0, length);
+            return new ReadOnlyCollection<string>(NullNotEmpty(str.Split(NameSeparator)));
         }
 
-        public DynamicFlagsCustomTypeInfo ToDynamicFlagsCustomTypeInfo()
+        private static byte[] CopyBytes(ReadOnlyCollection<byte> bytes, int start, int length)
         {
-            return DynamicFlagsCustomTypeInfo.Create(this);
+            var array = new byte[length];
+            for (int i = 0; i < length; i++)
+            {
+                array[i] = bytes[start + i];
+            }
+            return array;
         }
 
-        public DkmClrCustomTypeInfo ToDkmClrCustomTypeInfo()
+        private static string[] NullNotEmpty(string[] names)
         {
-            return Payload == null
-                ? null
-                : DkmClrCustomTypeInfo.Create(PayloadTypeId, Payload);
+            var builder = ArrayBuilder<string>.GetInstance(names.Length);
+            bool any = false;
+            foreach (var name in names)
+            {
+                if (name.Length == 0)
```

Comment:
Never mind, I found where it is used (`string.Split` above)

---
In reply to: [77422638](https://github.com/dotnet/roslyn/pull/13540#discussion_r77422638) [](ancestors = 77422638)